hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/CreateDagVisitor.java

```diff
@@ -529,36 +529,48 @@ public Vertex onUnion(UnionPhysicalRel rel) {
         return merger;
     }
 
-    public Vertex onRoot(RootRel rootRel) {
-        RelNode input = rootRel.getInput();
+    public Vertex onLimit(LimitPhysicalRel rel) {
+        RelNode input = rel.getInput();
+
         Expression<?> fetch;
         Expression<?> offset;
 
-        if (input instanceof SortPhysicalRel || isCalcWithSort(input)) {
-            SortPhysicalRel sortRel = input instanceof SortPhysicalRel
-                    ? (SortPhysicalRel) input
-                    : (SortPhysicalRel) ((CalcPhysicalRel) input).getInput();
+        if (rel.fetch() == null) {
+            fetch = ConstantExpression.create(Long.MAX_VALUE, QueryDataType.BIGINT);
+        } else {
+            fetch = rel.fetch(parameterMetadata);
+        }
 
-            if (sortRel.fetch == null) {
-                fetch = ConstantExpression.create(Long.MAX_VALUE, QueryDataType.BIGINT);
-            } else {
-                fetch = sortRel.fetch(parameterMetadata);
-            }
+        if (rel.offset() == null) {
+            offset = ConstantExpression.create(0L, QueryDataType.BIGINT);
+        } else {
+            offset = rel.offset(parameterMetadata);
+        }
 
-            if (sortRel.offset == null) {
-                offset = ConstantExpression.create(0L, QueryDataType.BIGINT);
-            } else {
-                offset = sortRel.offset(parameterMetadata);
-            }
+        Vertex vertex = dag.newUniqueVertex(
+                "LimitedClientSink",
+                rootResultConsumerSink(localMemberAddress, fetch, offset)
```

Comment:
This is wrong. The method `onLimit` can be called also in case there's a non-top-level LIMIT clause. In that case it would produce a 2nd client sink, which is wrong. 

The `onLimit` implementation here must throw, because we don't support LIMIT anywhere. We call the `onLimit` method from `onRoot`, but instead we should inline the current code into `onRoot`, because that code is only valid if the `LimitRel` is an input of `RootRel`.

Reply:
> The method onLimit can be called also in case there's a **non-top-level** LIMIT clause.

Calcite doesn't support multiple top-level LIMIT clauses, this is a point of current implementation.

Reply:
Try this
```
    @Test
    public void multipleLimitClauses() {
        String tableName = createTable(
                new String[]{"Alice", "1"},
                new String[]{"Bob", "2"},
                new String[]{"Joey", "3"}
        );

        assertThatThrownBy(() ->
                sqlService.execute(
                        " (SELECT distance FROM " + tableName + " LIMIT 2) " +
                                " UNION " +
                                " (SELECT distance FROM " + tableName + " LIMIT 2)"
                )).hasMessageContaining("FETCH/OFFSET is only supported for the top-level SELECT");
    }
```

Reply:
I know. But the code is just wrong. `onLimit` is called on LimitRel anywhere. If we leave it in a wrong place just because it doesn't happen today, it's a mess of convoluted code. And in the future when it's no longer true, you get very weird errors. I suspect, in this case, there even won't be an error, but the client sink will just cut off the rest of the DAG, the sink processor doesn't forward the items, but it will have downstream processors attached.

If some code works only in certain cases, at the very least we should add an assertion checking for those cases. But in this case it's better to move it to `onRoot`, because that's what the code does - it implements the root, which also does result limiting.

Reply:
Actually, the code was changed at the moment of writing it, checkout new commits :) The only thing I wanted to highlight -- we don't support multiple top-level LIMIT clauses, and that's it